Specs

Olsyx opened the pull request:
These are all the specs so far. 
They have been paralleled with the collision_improvement branch for proper testing of all the collision improvement on the game.

Olsyx committed:
code_refactoring

Olsyx committed:
solid_specs

Olsyx committed:
Ball issue fixed

Specs were not running properly due to the ball sprite's path being
taken from the entry point with webpack.
Issue has been fixed by commenting the line (for now), since the sprites
are not needed for the tests.

Olsyx committed:
specs

Specs for Canvanoid and Board

Olsyx committed:
specs

Added specs for board and solid, testing new collision system.

Commit:
Merge webpack.config.js into one

Olsyx committed:
Merge webpack.config.js into one

Commit:
Merge branch 'specs' of https://github.com/Darkatom/Canvanoid into specs

Olsyx committed:
Karma & Chai

Added Karma and Chai to the testing.

Olsyx committed:
Sinon

Added sinon

brauliodiez commented on index.html:
```diff
@@ -3,20 +3,11 @@
-		var game = new Canvanoid(canvas); 	
-		game.start();
-	</script>
+	<script src="./dist/canvanoid.bundle.js"></script> 
```
there is a plugin in webpack that already injects this in the html

 new HtmlWebpackPlugin({
      filename: 'index.html', //Name of file in ./dist/
      template: 'index.html', //Name of template in ./src
            hash: true
    })

Sample:

https://github.com/Lemoncode/webpack-1.x-by-sample/tree/master/00%20Intro/03%20Output

sirikon replied:
@brauliodiez Isn't better just leaving the script reference (which is something that rarely will change) instead of adding a whole new dependency? By simplicity

brauliodiez replied:
The thing is: you may want to add a hash to add browser caching, or maybe i. The future you just Split the thing in several bundles, e.g. app and vendor (by doing this if you make an small update on your app code but no updates on libraries you don't need to make the user Download the whole thing).

sirikon replied:
The hash for browser caching is just for development, and I didn't used that in a long time, becasue I already have the checkbox for disabling cache in Chrome Dev Tools.
Still the modifications that the index.html file could receive are minimal, still not convinced of using a whole new dependency for this.

Maybe I'm just tired of over-dependency :P

brauliodiez replied:
mmm, I use it the other way around, when you deploy your site the bundle js file will be kept in the browser cache, rather add some cache breaking mechanism to your bundle.js since you don't have control over the browser or a proxy server caching you js file.

sirikon replied:
In that case yes, would be necesary and useful.

Does HtmlWebpackPlugin perform a real hash of the file or it's just a random one?

brauliodiez replied:
Not sure, ideally it should be based on file content (like ASP .net mvc bundling does), but I don't know it.

brauliodiez replied:
Mmmm... is something easy to test, just run a couple of wepback generating to hard disk and check that the hash doesn't change because the content is the same, I will give a try and come back to you.

sirikon replied:
Checked the README of the plugin and they say that the hash changes 'every time you make a new compilation' Not clear what it does.

brauliodiez replied:
Ooohhh, well it would take some ticks plus whatever combination, but maybe we have a good chance to open a pull request on that project file CRC hash based should be the standard.

brauliodiez replied:
We can try to open an issue and check what they think, not sure how easy can be to implement this on a webpack plugin, but quite interesting.

sirikon replied:
Looking at the code, It seems like it takes the hash from the webpack compilaton, which uses this: https://webpack.github.io/docs/long-term-caching.html

brauliodiez commented on src/Canvanoid.js:
```diff
@@ -17,39 +17,93 @@ export default class Canvanoid {
 	
 	start() {	
+		// Game Elements
 		this.state = new State();
```
How many tabs are we using? ideally it should be only spaces and two spaces.

Olsyx replied:
Sirikon located an error during his previous revision for which GitHub would add extra spaces to the code for no apparent reason. He theorized that my tabbing set up might be somehow mixing up tabs and spaces, and that's definitely something I must look after.
I usually use four spaces (one tab) because, as far as I know, that's the standard, but I have no problem using two spaces instead.
It seems it happened again, so I will look more in depth into it. Last time it happened only to two lines, but I'm seeing here it just went crazy.
Thank you for noticing!

Olsyx replied:
@brauliodiez Fixed this, 2 spaces now.

brauliodiez commented on src/Canvanoid.js:
```diff
@@ -17,39 +17,93 @@ export default class Canvanoid {
-		this.vaus.start();
+		this.balls = [ new Ball(this.board.position.x + this.board.width/2, this.board.position.y + this.board.height/2 + 100) ];
+		this.vaus = new Paddle(this.board.position.x + this.board.width/2 - 50, 
+              				   this.board.position.y + this.board.height - 50);
```
We should follow clean code here, the code is hard to read, it would be easier to read something like

this.balls = createBallStartPosition();
this.vaus = createPaddleStartPosition();

And Create default ball contains all the plumbing

Olsyx replied:
@brauliodiez Fixed this adding two functions: one for putting the game objects in their initial positions, and another for the panels. I chose this way because the game objects (balls and vaus) are always reset together, so I figured I could just have a function for both of them. Same for the panels.

brauliodiez commented on src/Canvanoid.js:
```diff
@@ -17,39 +17,93 @@ export default class Canvanoid {
+		this.balls = [ new Ball(this.board.position.x + this.board.width/2, this.board.position.y + this.board.height/2 + 100) ];
+		this.vaus = new Paddle(this.board.position.x + this.board.width/2 - 50, 
+              				   this.board.position.y + this.board.height - 50);
+		this.interruptions();
```
What is the purpose of "interruptions" ?

Olsyx replied:
It's a function that initializes all the functions that take care of handling mouse and keyboard interruptions. In essence, the interaction of the user with the game.
Perhaps I should use a more intuitive name for it? I will try to come up with something.

brauliodiez replied:
Just thinking loud, some ideas:

```
setupUserInteractionEvents? 
```

```
setupEvents() {
   setupMouseEvents();
   setupKeyboardEvents();
}
```

"interruptions" sounds quite low level, the time when we were playing with int 33 (mouse handling) :)

Olsyx replied:
Hahaha, sorry, I picked up the "interrumptions" bit from a hardware subject in college. We programmed a small DS game, and I guess it stuck.
I like the setupEvents function! I will go with that one.

Olsyx replied:
@brauliodiez I fixed this with the setupEvents function.

brauliodiez commented on src/Canvanoid.js:
```diff
@@ -17,39 +17,93 @@ export default class Canvanoid {
 		
-		this.messagePanel = new Panel( this.board.x + this.board.w/2 - 150, 
-							           this.board.y + this.board.h/2 - 100);
+		this.instructionsPanel = new Panel(this.messagePanel.initialPosition.x, this.messagePanel.initialPosition.y + 50,
```
Same here I would encapsulate this in a function (and maybe all this I would isolate it in a different file (module) or class).

There is a principle called: Single level of abstraction level, you should read this code like a book, and it should have the same level of abstraction, not going deeper into coordinates and offsets

http://principles-wiki.net/principles:single_level_of_abstraction

brauliodiez commented on src/Canvanoid.js:
```diff
@@ -17,39 +17,93 @@ export default class Canvanoid {
-							           this.board.y + this.board.h/2 - 100);
+		this.instructionsPanel = new Panel(this.messagePanel.initialPosition.x, this.messagePanel.initialPosition.y + 50,
+							         	   this.ctx);
+		this.instructionsPanel.setSize("18");
```
Please don't use harcoded consts

Olsyx replied:
I'm not sure of how not to hardcode these consts.

I could make global variables (e.g.: _small_font_size, _big_font_size...) and use them to replace the "18" and other similar consts. However, they would still be passed to the function the same way -although I get why it makes the code more comprehensible. Would that be okay? Were you thinking of something else?

On this topic, the panels are often aligned left, center or right when needed. These are strings passed into the function, which causes the panel to rearrange itself. Would we consider them ("left", "center") hardcoded consts?

Olsyx replied:
@brauliodiez I added the global variables as indicated in my previous comment.

Sirikon commented on src/specs/src/board.spec.js:
```diff
@@ -0,0 +1,168 @@
+
+var expect = require('chai').expect;
+var should = require('chai').should();
+var assert = require('assert');
```
@Darkatom Should check tomorrow if we can use just import syntax, instead of both imports and requires :P

sirikon commented on src/specs/src/board.spec.js:
```diff
@@ -0,0 +1,168 @@
+
+var expect = require('chai').expect;
+var should = require('chai').should();
+var assert = require('assert');
```
@Darkatom Should check tomorrow if we can use just import syntax, instead of both imports and requires :P

brauliodiez replied:
import {expect} from 'chai'

Should do the trick, but not sure if 'chai' has something strange

Olsyx replied:
@brauliodiez, @Sirikon, I have fixed this for the expect require. However, it does not seem to work for _should()_. I will keep looking into it.

Olsyx committed:
Code refactoring

- Got rid of hardcoded consts.
- Created starting methods for game objects and panels.
- Fixed tabs.
- Other syntax fixes.

Olsyx committed:
Events fixed

Olsyx committed:
Tab fix

Olsyx committed:
Fix tabs

Olsyx reviewed (commented):
Comments on the last commit, trying to fix those issues pointed out on the last review.
